stellar/controllers/stellar_withdrawal_handler.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/go/support/render/problem"
+	"github.com/stellar/starbridge/backend"
+	"github.com/stellar/starbridge/ethereum"
+	"github.com/stellar/starbridge/store"
+)
+
+var validTxHash = regexp.MustCompile("^(0x)?([A-Fa-f0-9]{64})$")
+
+type StellarWithdrawalHandler struct {
+	StellarClient              *horizonclient.Client
+	Observer                   ethereum.Observer
+	Store                      *store.DB
+	StellarWithdrawalValidator backend.StellarWithdrawalValidator
+	EthereumFinalityBuffer     uint64
+}
+
+func (c *StellarWithdrawalHandler) getDeposit(r *http.Request, w http.ResponseWriter) (store.EthereumDeposit, error) {
+	txHash := r.PostFormValue("transaction_hash")
+	if !validTxHash.MatchString(txHash) {
+		w.WriteHeader(http.StatusBadRequest)
+		return store.EthereumDeposit{}, fmt.Errorf("invalid transaction hash")
+	}
+	parsed, err := strconv.ParseInt(r.PostFormValue("log_index"), 10, 32)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return store.EthereumDeposit{}, fmt.Errorf("invalid log index")
+	}
+	logIndex := uint(parsed)
+	depositID := ethereum.DepositID(txHash, logIndex)
+
+	storeDeposit, err := c.Store.GetEthereumDeposit(r.Context(), depositID)
+	if err == nil {
+		return storeDeposit, nil
+	} else if err != sql.ErrNoRows {
+		w.WriteHeader(http.StatusInternalServerError)
+		return store.EthereumDeposit{}, err
+	}
+
+	deposit, err := c.Observer.GetDeposit(r.Context(), txHash, logIndex)
+	if ethereum.IsInvalidGetDepositRequest(err) {
+		w.WriteHeader(http.StatusBadRequest)
+		return store.EthereumDeposit{}, fmt.Errorf("invalid log index")
+	} else if err == ethereum.ErrTxHashNotFound {
+		w.WriteHeader(http.StatusNotFound)
+		return store.EthereumDeposit{}, err
+	} else if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return store.EthereumDeposit{}, err
+	}
+
+	block, err := c.Observer.GetLatestBlock(r.Context())
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return store.EthereumDeposit{}, err
+	}
+	if deposit.BlockNumber+c.EthereumFinalityBuffer > block.Number {
+		w.WriteHeader(http.StatusPreconditionFailed)
+		return store.EthereumDeposit{}, fmt.Errorf("need to wait for finality buffer")
+	}
+
+	destinationAccountID, err := strkey.Encode(
+		strkey.VersionByteAccountID,
+		deposit.Destination.Bytes(),
+	)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return store.EthereumDeposit{}, fmt.Errorf("invalid stellar destination account %w", err)
+	}
+	storeDeposit = store.EthereumDeposit{
+		ID:          depositID,
+		Token:       deposit.Token.String(),
+		Sender:      deposit.Sender.String(),
+		Destination: destinationAccountID,
+		Amount:      deposit.Amount.String(),
+		Hash:        deposit.TxHash.String(),
+		LogIndex:    deposit.LogIndex,
+		BlockNumber: deposit.BlockNumber,
+		Timestamp:   deposit.Time.Unix(),
+	}
+	if err = c.Store.InsertEthereumDeposit(r.Context(), storeDeposit); err != nil {
```

Comment:
Should we move inserting new deposits to backend worker? I'm afraid we're exposing eth client to the public doing it here. Not sure if it's safe.

Reply:
this is a good point but if the controller is compromised it can potentially modify the db so I'm not sure if there is much security benefit to only insert ethereum deposits in the backend worker. it would be good to for us to figure out what privileges the controller should and shouldn't have compared to the backend worker

Reply:
I was thinking more about DoS scenarios in which somebody sends a bunch of random deposit IDs.

Reply:
I think it's still possible to do that even if we move the logic to the backend worker. the only difference would be that the eth client calls would be moved to the worker instead of the http handler